README.md
# ES Module Shims

Shims modern ES Modules features like import maps on top of the baseline modules support in browsers supported by [95% of users](https://caniuse.com/#feat=es6-module).

When running in polyfill mode, [the 67% of users](https://caniuse.com/import-maps) with import maps entirely bypass the shim code entirely.

For the remaining 30% of users, the highly performant (see [benchmarks](#benchmarks)) production and [CSP-compatible](#csp-support) shim kicks in to rewrite module specifiers driven by the [Web Assembly ES Module Lexer](https://github.com/guybedford/es-module-lexer).

The following modules features are polyfilled:

* [Import Maps](#import-maps) support.
* Dynamic `import()` shimming when necessary in eg older Firefox versions.
* `import.meta` and `import.meta.url`.
* [JSON](#json-modules) and [CSS modules](#css-modules) with import assertions (when enabled).
* [`<link rel="modulepreload">` polyfill](#modulepreload) in non Chromium browsers for both shimmed and unshimmed preloading scenarios.

When running in shim mode, module rewriting is applied for all users and custom [resolve](#resolve-hook) and [fetch](#fetch-hook) hooks can be implemented allowing for custom resolution and streaming in-browser transform workflows.

Because we are still using the native module loader the edge cases work out comprehensively, including:

* Live bindings in ES modules
* Dynamic import expressions (`import('src/' + varname')`)
* Circular references, with the execption that live bindings are disabled for the first unexecuted circular parent.

> [Built with](https://github.com/guybedford/es-module-shims/blob/main/chompfile.toml) [Chomp](https://chompbuild.com/)

## Usage

Include ES Module Shims with a `async` attribute on the script, then include an import map and module scripts normally:

```html
<script async src="https://ga.jspm.io/npm:es-module-shims@1.5.2/dist/es-module-shims.js"></script>

<!-- https://generator.jspm.io/#U2NhYGBkDM0rySzJSU1hKEpNTC5xMLTQM9Az0C1K1jMAAKFS5w0gAA -->
<script type="importmap">
{
  "imports": {
    "react": "https://ga.jspm.io/npm:react@18.0.0-rc.0/index.js"
  },
  "scopes": {
    "https://ga.jspm.io/npm:react@18.0.0-rc.0/": {
      "object-assign": "https://ga.jspm.io/npm:object-assign@4.1.1/index.js"
    }
  }
}
</script>

<script type="module">
  import react from 'react';
  console.log(react);
</script>
```

In browsers without import maps support, a console error will be given:

```
Uncaught TypeError: Failed to resolve module specifier "app". Relative references must start with either "/", "./", or "../".
  at <anonymous>:1:15
```

This execution failure is a feature - it avoids the polyfill causing double execution. The first import being a bare specifier in the pattern above is important to ensure this.

This is because the polyfill cannot disable the native loader - instead it will only execute modules that would otherwise fail resolving or parsing to avoid duplicate fetches or executions that would cause performance and reliability issues.

### Shim Mode

Shim mode is an alternative to polyfill mode and doesn't rely on native modules erroring - instead it is triggered by the existence of any `<script type="importmap-shim">` or `<script type="module-shim">`, or when explicitly setting the [`shimMode` init option](#shim-mode-option).

In shim mode, only the above `importmap-shim` and `module-shim` tags will be parsed and executed by ES Module Shims.

Shim mode also provides some additional features that aren't yet natively supported such as supporting multiple import maps, [external import maps](#external-import-maps) with a `"src"` attribute, [dynamically injecting import maps](#dynamic-import-maps), and [reading current import map state](#reading-current-import-map-state), which can be useful in certain applications.

## Benchmarks

ES Module Shims is designed for production performance. A [comprehensive benchmark suite](bench/README.md) tracks multiple loading scenarios for the project.

Benchmark summary:

* [ES Module Shims Chrome Passthrough](bench/README.md#chrome-passthrough-performance) (for [70% of users](https://caniuse.com/import-maps)) results in ~5ms extra initialization time over native for ES Module Shims fetching, execution and initialization, and on a slow connection the additional non-blocking bandwidth cost of its 10KB compressed download as expected.
* [ES Module Shims Polyfilling](bench/README.md#native-v-polyfill-performance) (for the remaining [30% of users](https://caniuse.com/import-maps)) is on average 1.4x - 1.5x slower than native module loading, and up to 1.8x slower on slow networks (most likely due to the browser preloader), both for cached and uncached loads, and this result scales linearly up to 10MB and 20k modules loaded executing on the fastest connection in just over 2 seconds in Firefox.
* [Very large import maps](bench/README.md#large-import-maps-performance) (100s of entries) cost only a few extra milliseconds upfront for the additional loading cost.

## Features

### Browser Support

Works in all browsers with [baseline ES module support](https://caniuse.com/#feat=es6-module).

Browser Compatibility **with ES Module Shims**:

| ES Modules Features                | Chrome (61+)                         | Firefox (60+)                        | Safari (10.1+)                       | Edge (17+)                           |
| ---------------------------------- | ------------------------------------ | ------------------------------------ | ------------------------------------ | ------------------------------------ |
| Executes Modules in Correct Order  | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:<sup>1</sup>       |
| [Dynamic Import](#dynamic-import)  | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   |
| [import.meta.url](#importmetaurl)  | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   |
| [Module Workers](#module-workers)  | :heavy_check_mark: ~68+              | :x:<sup>2</sup>                      | :x:<sup>2</sup>                      | :x:<sup>2</sup>                      |
| [modulepreload](#modulepreload)    | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   |
| [Import Maps](#import-maps)        | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   |
| [JSON Modules](#json-modules)      | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   |
| [CSS Modules](#css-modules)        | :heavy_check_mark:<sup>3</sup>       | :heavy_check_mark:<sup>3</sup>       | :heavy_check_mark:<sup>3</sup>       | :heavy_check_mark:<sup>3</sup>       |
| [import.meta.resolve](#resolve)    | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   |
| [Top-Level Await](#tla)            | :heavy_check_mark: 89+               | :heavy_check_mark: 89+               | :x:<sup>4</sup>                      | :x:<sup>4</sup>                       |

* 1: _The Edge parallel execution ordering bug is corrected by ES Module Shims with an execution chain inlining approach._
* 2: _Module worker support cannot be implemented without dynamic import support in web workers._
* 3: _CSS module support requires a separate [Constructable Stylesheets polyfill](https://github.com/calebdwilliams/construct-style-sheets#readme)._
* 4: _Top-level await support is possible for ES Module Shims to implement, with the feature request tracking in https://github.com/guybedford/es-module-shims/issues/5._

Browser compatibility **without ES module shims**:

| ES Modules Features                | Chrome (61+)                         | Firefox (60+)                        | Safari (10.1+)                       | Edge (17+)                           |
| ---------------------------------- | ------------------------------------ | ------------------------------------ | ------------------------------------ | ------------------------------------ |
| Executes Modules in Correct Order  | :heavy_check_mark:                   | :heavy_check_mark:                   | :heavy_check_mark:                   | :x:<sup>1</sup>                      |
| [Dynamic Import](#dynamic-import)  | :heavy_check_mark: 63+               | :heavy_check_mark: 67+               | :heavy_check_mark: 11.1+             | :x:                                  |
| [import.meta.url](#importmetaurl)  | :heavy_check_mark: ~76+              | :heavy_check_mark: ~67+              | :heavy_check_mark: ~12+ ❕<sup>1</sup>| :x:                                  |
| [Module Workers](#module-workers)  | :heavy_check_mark: ~68+              | :x:                                  | :x:                                  | :x:                                  |
| [modulepreload](#modulepreload)    | :heavy_check_mark: 66+               | :x:                                  | :x:                                  | :x:                                  |
| [Import Maps](#import-maps)        | :heavy_check_mark: 89+               | :x:                                  | :x:                                  | :x:                                  |
| [JSON Modules](#json-modules)      | :heavy_check_mark: 91+               | :x:                                  | :x:                                  | :x:                                  |
| [CSS Modules](#css-modules)        | :heavy_check_mark: 95+               | :x:                                  | :x:                                  | :x:                                  |
| [import.meta.resolve](#resolve)    | :x:                                  | :x:                                  | :x:                                  | :x:                                  |
| [Top-Level Await](#tla)            | :heavy_check_mark: 89+               | :heavy_check_mark: 89+               | :x:                                  | :x:                                  |

* 1: _Edge executes parallel dependencies in non-deterministic order. ([ChakraCore bug](https://github.com/microsoft/ChakraCore/issues/6261))._
* ~: _Indicates the exact first version support has not yet been determined (PR's welcome!)._
* ❕<sup>1</sup>: On module redirects, Safari returns the request URL in `import.meta.url` instead of the response URL as per the spec.

### Import Maps

> Stability: WhatWG Standard, Single Browser Implementer

Import maps allow for importing "bare specifiers" in JavaScript modules, which prior to import maps throw in all browsers with native modules support.

Using this polyfill we can write:

```html
<script type="importmap-shim">
{
  "imports": {
    "test": "/test.js"
  },
  "scopes": {
    "/": {
      "test-dep": "/test-dep.js"
    }
  }
}
</script>
<script type="module-shim">
  import test from "test";
  console.log(test);
</script>
```

All modules are still loaded with the native browser module loader, but with their specifiers rewritten then executed as Blob URLs, so there is a relatively minimal overhead to using a polyfill approach like this.

#### Multiple Import Maps

Multiple import maps are not currently supported in any native implementation, Chromium support is currently being tracked in https://bugs.chromium.org/p/chromium/issues/detail?id=927119.

In polyfill mode, multiple import maps are therefore not supported.

In shim mode, support for multiple `importmap-shim` scripts follows the [import map extensions](https://github.com/guybedford/import-maps-extensions) proposal.

#### External Import Maps

External import maps (using a `"src"` attribute) are not currently supported in any native implementation.

In polyfill mode, external import maps are therefore not supported.

In shim mode, external import maps are fully supported.

#### Dynamic Import Maps

Support for dynamically injecting import maps with JavaScript via eg:

```js
document.body.appendChild(Object.assign(document.createElement('script'), {
  type: 'importmap',
  innerHTML: JSON.stringify({ imports: { x: './y.js' } }),
}));
```

is supported in Chromium, provided it is injected before any module loads and there is no other import map yet loaded (multiple import maps are not supported).

Both modes in ES Module Shims support dynamic injection using DOM Mutation Observers.

While in polyfill mode the same restrictions apply that multiple import maps, import maps with a `src` attribute, and import maps loaded after the first module load are not supported, in shim mode all of these behaviours are fully enabled for `"importmap-shim"`.

#### Reading current import map state

To make it easy to keep track of import map state, es-module-shims provides a `importShim.getImportMap` utility function, available only in shim mode.

```js
const importMap = importShim.getImportMap()

// importMap will be an object in the same shape as the json in a importmap script
```

### Dynamic Import

> Stability: Stable browser standard

Dynamic `import(...)` within any modules loaded will be rewritten as `importShim(...)` automatically
providing full support for all es-module-shims features through dynamic import.

To load code dynamically (say from the browser console), `importShim` can be called similarly:

```js
importShim('/path/to/module.js').then(x => console.log(x));
```

### import.meta.url

> Stability: Stable browser standard

`import.meta.url` provides the full URL of the current module within the context of the module execution.

### modulepreload

> Stability: WhatWG Standard, Single Browser Implementer

Preloading of modules can be achieved by including a `<link rel="modulepreload" href="/module.js" />` tag in the HTML or injecting it dynamically.

This tag also supports the `"integrity"`, `"crossorigin"` and `"referrerpolicy"` attributes as supported on module scripts.

This tag just initiates a fetch request in the browser and thus works equally as a preload polyfill in both shimmed and unshimmed modes, with integrity validation support.

Unlike the browser specification, the modulepreload polyfill does not request dependency modules by default, in order to avoid unnecessary
code analysis in the polyfill scenarios. **It is recommended to preload deep imports anyway so that this feature shouldn't be necessary.**

#### Preload shim

When in shim mode, `<link rel="modulepreload-shim" href="/module.js" />` must be used to properly cache the preloaded modules.

### CSP Support

By default ES Module Shims provides full support for CSP by using the asm.js ES Module Lexer build. This is absolutely identical in performance to the Wasm version in Firefox and Chrome, while in Safari the asm.js version is actually faster than Wasm making this build preferable.

The CSP nonce to use for module scripts will be picked up from the first script on the page or via the [`nonce` init option](#nonce).

A full example of such a CSP workflow is provided below:

```html
<meta http-equiv="Content-Security-Policy" content="script-src 'self' 'nonce-n0nce'" />
<script async src="es-module-shims.js"></script>
<script type="importmap" nonce="n0nce">
{
  "pkg": "/pkg.js"
}
</script>
<script type="module" nonce="n0nce">
import pkg from 'pkg';
</script>
```

#### Wasm Build

To use the Web Assembly / non-CSP build of ES Module Shims, this is available as a self-contained single file at `es-module-shims/wasm` or `es-module-shims/dist/es-module-shims.wasm.js` in the package folder.

### JSON Modules

> Stability: WhatWG Standard, Single Browser Implementer

In shim mode, JSON modules are always supported. In polyfill mode, JSON modules require the `polyfillEnable: ['json-modules']` [init option](#polyfill-enable-option).

JSON Modules are currently supported in Chrome when using them via an import assertion:

```html
<script type="module">
import json from 'https://site.com/data.json' assert { type: 'json' };
</script>
```

In addition JSON modules need to be served with a valid JSON content type.

Checks for assertion failures are not currently included.

### CSS Modules

> Stability: WhatWG Standard, Single Browser Implementer

In shim mode, CSS modules are always supported. In polyfill mode, CSS modules require the `polyfillEnable: ['css-modules']` [init option](#polyfill-enable-option).

CSS Modules are currently supported in Chrome when using them via an import assertion:

```html
<script type="module">
import sheet from 'https://site.com/sheet.css' assert { type: 'css' };
</script>
```

To support the polyfill or shim of this feature, the [Constructable Stylesheets polyfill](https://github.com/calebdwilliams/construct-style-sheets#readme) must be separately included in browsers not supporting [Constructable Stylesheets](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/CSSStyleSheet) eg via:

```html
<script async src="https://unpkg.com/construct-style-sheets-polyfill@3.1.0/dist/adoptedStyleSheets.js"></script>
```

For more information see the [web.dev article](https://web.dev/css-module-scripts/).

In addition CSS modules need to be served with a valid CSS content type.

Checks for assertion failures are not currently included.

### Resolve

> Stability: Draft HTML PR

`import.meta.resolve` provides a contextual resolver within modules. It is synchronous, changed from being formerly asynchronous due to following the [browser specification PR](https://github.com/whatwg/html/pull/5572).

The second argument to `import.meta.resolve` permits a custom parent URL scope for the resolution (not currently in the browser spec), which defaults to `import.meta.url`.

```js
// resolve a relative path to a module
var resolvedUrl = import.meta.resolve('./relative.js');
// resolve a dependency from a module
var resolvedUrl = import.meta.resolve('dep');
// resolve a path
var resolvedUrlPath = import.meta.resolve('dep/');
// resolve with a custom parent scope
var resolvedUrl = import.meta.resolve('dep', 'https://site.com/another/scope');
```

Node.js also implements a similar API, although it's in the process of shifting to a synchronous resolver.

## Polyfill Mode Details

In polyfill mode, feature detections are performed for ES modules features. In browsers with full feature support no further processing is done.

In browsers with variable feature support, sources are analyzed with module specifiers rewritten using the very fast Wasm / asm.js lexer while sharing the source network fetch cache with the native loader.

#### Polyfill Features

The current default native baseline for the ES module shims polyfill mode is browsers supporting import maps.

If using more modern features like CSS Modules or JSON Modules, these need to be manually enabled via the [`polyfillEnable` init option](#polyfill-enable-option) to raise the native baseline to only browsers supporting these features:

```html
<script>
window.esmsInitOptions = { polyfillEnable: ['css-modules', 'json-modules'] }
</script>
```

To verify when the polyfill is actively engaging as opposed to relying on the native loader, [a `polyfill` hook](#polyfill-hook) is also provided.

#### Polyfill Edge Case: Dynamic Import

The guarantee of the polyfill is that any module graph that would have failed will be reexecuted through the shim layer. This leaves any edge case where execution succeeds but not as expected. For example when using dynamic imports:

```html
<script type="module">
  console.log('Executing');
  const dynamic = 'bare-specifier';
  import(dynamic).then(x => {
    console.log('Ok');
  }, err => {
    console.log('Fail');
  });
</script>
```

The native browser loader without import maps support will execute the above module fine, but fail on the lazy dynamic import.

ES Module Shims will not reexecute the above in browsers without import maps support though because it will see that the execution did complete successfully therefore it will not attempt reexecution and as a result, `"Ok"` is never logged.

Other examples include dynamically injecting import maps, or using import maps with a `"src"` attribute, which aren't supported in native Chrome.

This is why it is advisable to always ensure modules use syntax that will fail early to avoid non-execution.

#### Polyfill Edge Case: Instance Sharing

When running in polyfill mode, it can be thought of that are effectively two loaders running on the page - the ES Module Shims polyfill loader, and the native loader.

Note that instances are not shared between these loaders for consistency and performance.

As a result, if you have two module graphs - one native and one polyfilled, they will not share the same dependency instance, for example:

```html
<script type="importmap">
{
  "imports": {
    "dep": "/dep.js"
  }
}
</script>
<script type="module">
import '/dep.js';
</script>
<script type="module">
import 'dep';
</script>
```

In the above, on browsers without import maps support, the `/dep.js` instance will be loaded natively by the first module, then the second import will fail.

ES Module Shims will pick up on the second import and reexecute `/dep.js`. As a result, `/dep.js` will be executed twice on the page.

For this reason it is important to always ensure all modules hit the polyfill path, either by having all graphs use import maps at the top-level, or via `importShim` directly.

#### Skip Polyfill

Adding the `"noshim"` attribute to the script tag will also ensure that ES Module Shims skips processing this script entirely:

```html
<script type="module" noshim>
  // ...
</script>
```

## Init Options

Provide a `esmsInitOptions` on the global scope before `es-module-shims` is loaded to configure various aspects of the module loading process:

* [shimMode](#shim-mode-option)
* [polyfillEnable](#polyfill-enable-option)
* [enforceIntegrity](#enforce-integrity)
* [nonce](#nonce)
* [noLoadEventRetriggers](#no-load-event-retriggers)
* [skip](#skip-processing)
* [onerror](#error-hook)
* [onpolyfill](#polyfill-hook)
* [resolve](#resolve-hook)
* [fetch](#fetch-hook)
* [revokeBlobURLs](#revoke-blob-urls)
* [mapOverrides](#overriding-import-map-entries)

```html

CHANGELOG.md
ES Module Shims 1.5.1 (2022/03/17)
* Deps: Update to es-module-lexer@0.10.4 (#273)

ES Module Shims 1.5.0 (2022/03/17)
* Feat: Import hook, meta hook, sync resolver implementation, deprecating async resolver (#268)
* Deps: Update to es-module-lexer@0.10.3 (#272)

ES Module Shims 1.4.7 (2022/02/23)
* Feat: Alternative IFrame injection to avoid navigation event on Android (#265)
* Deps: Update to es-module-lexer@0.10.0 (#266)

ES Module Shims 1.4.6 (2022/02/15)
* Feat: Improve fetch content type error failure context (https://github.com/guybedford/es-module-shims/pull/263)
* Feat: Refactor error handling paths, use `reportError` when possible (https://github.com/guybedford/es-module-shims/pull/264)

ES Module Shims 1.4.5 (2022/02/12)
* Feat: Allow import map overrides with option (https://github.com/guybedford/es-module-shims/pull/257, @lewisl@9029)
* Feat: Additional fetch error context (https://github.com/guybedford/es-module-shims/pull/256)
* Fix: dynamic import context with import assertions (https://github.com/guybedford/es-module-shims/pull/260, @ICodeMyOwnLife)
* Docs: Clarify modulepreload-shim (https://github.com/guybedford/es-module-shims/pull/261, @LarsDenBakker)

ES Module Shims 1.4.4 (2022/02/03)
* Fix: import.meta.resolve support for inline scripts (https://github.com/guybedford/es-module-shims/pull/251)
* Fix: CSS imports unquoted URL normalization (https://github.com/guybedford/es-module-shims/pull/252)
* Fix: Reword polyfill info message (https://github.com/guybedford/es-module-shims/pull/253)

ES Module Shims 1.4.3 (2022/01/25)
* Feature: Support import map overrides when existing resolution matches new resolution (https://github.com/guybedford/es-module-shims/pull/249, @vovacodes)
* Fix: Import assertions support with dynamic imports (https://github.com/guybedford/es-module-shims/pull/248)

ES Module Shims 1.4.2 (2022/01/24)
* Fix: optional catch binding in `isUrl` function (@thepassle, https://github.com/guybedford/es-module-shims/pull/246)

ES Module Shims 1.4.1 (2021/12/15)
* Fix: Firefox parse error stacks (https://github.com/guybedford/es-module-shims/pull/239)

ES Module Shims 1.4.0 (2021/12/14)
* Feature: "enforceIntegrity" option and "modulepreload-shim" support for secure shim mode with integrity (https://github.com/guybedford/es-module-shims/pull/236)
* Fix: Safari parse error stacks (https://github.com/guybedford/es-module-shims/pull/238)
* Note: Polyfill mode console note when expected native errors are present (https://github.com/guybedford/es-module-shims/pull/237)

ES Module Shims 1.3.6 (2021/12/07)
* Fix resolve hook via esms options (#233, @ffortier)

ES Module Shims 1.3.5 (2021/12/02)
* Fix trailing comment regex for source mapping normalization with trailling comments (https://github.com/guybedford/es-module-shims/pull/231)

ES Module Shims 1.3.4 (2021/12/02)
* Support source mapping normalization with trailing comments (https://github.com/guybedford/es-module-shims/pull/229)

ES Module Shims 1.3.3 (2021/11/30))
* Remove use of document.write (https://github.com/guybedford/es-module-shims/pull/227, @lewisl9029)

ES Module Shims 1.3.2 (2021/11/16)
* Fix CSP nonce detection (https://github.com/guybedford/es-module-shims/pull/223, @MathiasWP)

ES Module Shims 1.3.1 (2021/11/04)
* Feature: Support "async" attribute ordering on polyfill and shim scripts (https://github.com/guybedford/es-module-shims/pull/221)
* Fix: Sourcemap regex fix (https://github.com/guybedford/es-module-shims/pull/218)
* Fix: Polyfill duplicate instance bug (https://github.com/guybedford/es-module-shims/pull/220)

ES Module Shims 1.3.0 (2021/10/14)
* Feature: Shim mode function to get import map, dynamic map overrides now throw (https://github.com/guybedford/es-module-shims/pull/213, @lewisl9029)
* Feature: Upgrade to es-module-lexer@0.9.3 with import / export strings support (https://github.com/guybedford/es-module-shims/pull/210)
* Performance: Disable subgraph passthrough (https://github.com/guybedford/es-module-shims/pull/208)
* Fix: Undefined error callbacks for Safari (https://github.com/guybedford/es-module-shims/pull/209)

ES Module Shims 1.2.0 (2021/10/04)
* Fix: Bug where fetch hook was renamed to "fetchHook" instead of "fetch" (https://github.com/guybedford/es-module-shims/pull/205)

ES Modules Shims 1.1.0 (2021/09/28)
* Feature: Enable polyfill when loading a module before the import map. This resolves browser extension issues (https://github.com/WICG/import-maps/issues/248) (https://github.com/guybedford/es-module-shims/pull/203)
* Fix: Upgrade to es-module-lexer@0.9.1 supporting large source analysis (https://github.com/guybedford/es-module-shims/pull/203)

ES Module Shims 1.0.4 (2021/09/22)
* Fix typo bug in import maps call (https://github.com/guybedford/es-module-shims/pull/198)

ES Module Shims 1.0.3 (2021/09/22)
* Lifecycle event polishing and fixes (https://github.com/guybedford/es-module-shims/pull/197)

ES Module Shims 1.0.2 (2021/09/22)
* Fix native passthrough for contextual `importShim` dynamic import to still full reconstruct top-level resolution by parsing import maps (https://github.com/guybedford/es-module-shims/pull/196)
* Remove skip option default (https://github.com/guybedford/es-module-shims/pull/195)

ES Module Shims 1.0.1 (2021/09/22)
* Improved native passthrough performance, with some load event restrictions on unstable new features within release cutoff (https://github.com/guybedford/es-module-shims/pull/194)

ES Module Shims 1.0.0 (2021/09/20)
* CSP support build by default (https://github.com/guybedford/es-module-shims/pull/193)

ES Module Shims 0.15.1 (2021/09/18)
* Significant performance improvements to CSP build via asm.js (https://github.com/guybedford/es-module-shims/pull/190) and direct dynamic import (https://github.com/guybedford/es-module-shims/pull/191)

ES Module Shims 0.15.0 (2021/09/17)
* Various perf and behaviour polishing (https://github.com/guybedford/es-module-shims/pull/188)

bench/generate.js
import { readFile, writeFile, mkdir, readdir } from 'fs/promises';
import { join, sep } from 'path';
import rimraf from 'rimraf';

// GENERATE CASES
{
  const preact = await readFile(new URL('../node_modules/preact/dist/preact.module.js', import.meta.url));

  try {
    rimraf.sync(new URL('./generated', import.meta.url));
  }
  catch {}

  try {
    await mkdir(new URL('./generated', import.meta.url));
  }
  catch {}

  await writeFile(new URL('./generated/preact.js', import.meta.url), preact);

  for (let i = 1; i <= 10000; i++) {
    await writeFile(new URL(`./generated/app${i}.js`, import.meta.url), `\
  import { h, Component, render } from './preact.js?n=${i}';

  export class App extends Component {
    render() {
      return h('h1', null, 'Hello, world ${i}!');
    }
  }

  const el = document.createElement('div');
  render(h(App), el);
  `);
  }

  for (let i = 1; i <= 10000; i++) {
    await writeFile(new URL(`./generated/app.mapped${i}.js`, import.meta.url), `\
  import { h, Component, render } from 'lib/preact.js?n=${i}';

  export class App extends Component {
    render() {
      return h('h1', null, 'Hello, world ${i}!');
    }
  }

  const el = document.createElement('div');
  render(h(App), el);
  `);
  }

  // GENERATE ALLMAPPED TESTS
  for (const n of [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 1000]) {
    const imports = {};
    for (let i = 1; i <= n; i++) {
      imports[`lib/app.mapped${i}.js`] = `/bench/generated/app.mapped${i}.js`;
      imports[`lib/preact.js?n=${i}`] = `/bench/generated/preact.js?n=${i}`;
    }

    await writeFile(new URL(`./generated/parallel-allmapped-${n}.html`, import.meta.url), `\
<script type="importmap">
{
  "imports": ${JSON.stringify(imports)}
}
</script>
<script type="module">
  let promises = Array(${n});
  for (let i = 1; i <= ${n}; i++)
    promises.push(import(\`lib/app.mapped\${i}.js\`));
  Promise.all(promises).then(() => parent.postMessage('done', '*'));
</script>
`);

    await writeFile(new URL(`./generated/parallel-allmapped-esms-${n}.html`, import.meta.url), `\
<script type="importmap">
{
  "imports": ${JSON.stringify(imports)}
}
</script>
<script async src="/dist/es-module-shims.js"></script>
<script type="module">
  import 'lib/app.mapped1.js';
  let promises = Array(${n});
  for (let i = 1; i <= ${n}; i++)
    promises.push(import(\`lib/app.mapped\${i}.js\`));
  Promise.all(promises).then(() => parent.postMessage('done', '*'));
</script>
`);
  }
}

// GENERATE BENCHMARKS

{
  try { await mkdir(new URL('./results', import.meta.url)) }
  catch {}

  try { rimraf.sync(new URL('./benchmarks', import.meta.url)) }
  catch {}

  try { await mkdir(new URL('./benchmarks', import.meta.url)) }
  catch {}

  const ports = {
    'fastest': 8000,
    'fastest-cached': 8001,
    'slow': 8002,
    // 'slow-uncompressed': 8003
  };

  for (const browser of ['safari', 'firefox', 'chrome']) {
    let firefoxProfilePath;
    if (browser === 'firefox' && process.env.APPDATA) {
      // TODO: Firefox profile path for other platforms
      const profileDir = join(process.env.APPDATA, 'Mozilla', 'Firefox', 'Profiles');
      const profiles = await readdir(profileDir);
      firefoxProfilePath = profileDir + sep + profiles.sort()[0];
    }
    for (const type of Object.keys(ports)) {
      for (const name of [
        'parallel',
        'parallel-mapped-esms',
        'parallel-allmapped-esms',
        ...browser === 'chrome' ? ['parallel-mapped', 'parallel-allmapped'] : [],
      ]) {
        const fullName = `${browser}.${type}.${name}`;
        const port = ports[type];
        await writeFile(new URL(`./benchmarks/${fullName}.bench.json`, import.meta.url), `\
        {
          "$schema": "https://raw.githubusercontent.com/Polymer/tachometer/master/config.schema.json",
          "sampleSize": 20,
          "timeout": 0,
          "autoSampleConditions": ["0%"],
          "benchmarks": [
            {
              "name": "${fullName}-10",
              "url": "runner.html?type=${name}&n=10&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-20",
              "url": "runner.html?type=${name}&n=20&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-30",
              "url": "runner.html?type=${name}&n=30&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-40",
              "url": "runner.html?type=${name}&n=40&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-50",
              "url": "runner.html?type=${name}&n=50&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-60",
              "url": "runner.html?type=${name}&n=60&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-70",
              "url": "runner.html?type=${name}&n=70&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-80",
              "url": "runner.html?type=${name}&n=80&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-90",
              "url": "runner.html?type=${name}&n=90&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            },
            {
              "name": "${fullName}-100",
              "url": "runner.html?type=${name}&n=100&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            }${type.startsWith('slow') ? '' : `,
            {
              "name": "${fullName}-1000",
              "url": "runner.html?type=${name}&n=1000&port=${port}",
              "browser": {
                "name": "${browser}"${browser === 'firefox' ? `,
                "profile": ${JSON.stringify(firefoxProfilePath)}` : ''}
              }
            }`}
          ]
        }
        `);
      }
    }
  }
}
